src/main/kotlin/io/libp2p/core/Stream.kt

```diff
@@ -1,26 +1,69 @@
 package io.libp2p.core
 
-import io.libp2p.etc.PROTOCOL
-import io.netty.channel.Channel
+import com.google.protobuf.MessageLite
+import io.libp2p.protocol.ProtocolMessageHandler
+import io.libp2p.protocol.ProtocolMessageHandlerAdapter
+import io.netty.handler.codec.protobuf.ProtobufDecoder
+import io.netty.handler.codec.protobuf.ProtobufEncoder
+import io.netty.handler.codec.protobuf.ProtobufVarint32FrameDecoder
+import io.netty.handler.codec.protobuf.ProtobufVarint32LengthFieldPrepender
+import java.lang.reflect.ParameterizedType
 import java.util.concurrent.CompletableFuture
 
 /**
  * Represents a multiplexed stream over wire connection
  */
-class Stream(ch: Channel, val conn: Connection) : P2PAbstractChannel(ch) {
-
-    init {
-        nettyChannel.attr(PROTOCOL).set(CompletableFuture())
-    }
+interface Stream : P2PChannel {
+    val connection: Connection
 
     /**
      * Returns the [PeerId] of the remote peer [Connection] which this
      * [Stream] created on
      */
-    fun remotePeerId() = conn.secureSession.remoteId
+    fun remotePeerId(): PeerId
 
     /**
      * @return negotiated protocol
      */
-    fun getProtocol(): CompletableFuture<String> = nettyChannel.attr(PROTOCOL).get()
-}
\ No newline at end of file
+    fun getProtocol(): CompletableFuture<String>
+
+    fun <TMessage> pushHandler(protocolHandler: ProtocolMessageHandler<TMessage>) {
```

Comment:
Why do we deal with `Protobuf` in the `Stream` interface?

Reply:
What I'm trying to do in this PR is find the "right" place for methods to sit, to try make sure the interfaces have the operations people need. Given that the library is currently quite small, that might mean that some of the choices I've made will turn out to have been too restricting. In that case, we'll just have to make the appropriate changes as the need arises. To me, that always feels better, than going overly liberal. 

In libp2p terms, for example, Connection and Stream are different concepts, representing different things. That fact that we've implemented them using netty.Channel's isn't particularly important, and doesn't make them the same. Nor is it an argument for exposing the netty.Channel. What do, for example, Channel.parent() or Channel.config() mean in the context of a libp2p Stream? 

I pushed the Protobuf handling down into Stream because it seemed to me the best place for it. Again, that might turn out to be wrong in which case we can change it, but my thinking was that *any* protocol that uses a protobuf message needs to add the four Protobuf handlers before adding their own handler. Those four handlers therefore constitute boilerplate. We could document that - "if you use a protobuf, you need to do this" - but if we can document it then we can program it away. Part of the job of a library, surely, is to reduce boilerplate, so that's what I tried to do. Because of the type-erasure inherent in JVM generics, the implementation involves reflection and walking the type hierarchy but there's no way around that, unfortunately. The benefit, as I see it, is that the implementation of Identify, currently our only protobuf based protocol, is similar to Ping and the test protocols we've defined.

Reply:
Though the libp2p itself is heavily using protobuf the endpoint protocol is free to use any serialization from an unlimited set of possible options. Or for example it may use the protobuf but with alternative message delimiters. You shouldn't handle any possible option here. 

What do you think on this option:
https://github.com/libp2p/jvm-libp2p/commit/80add1550261c4bd9780e90f2f9003b792f3c8f9

Reply:
It still makes a special case of protobuf messages and requires users to know that they're a special case. With this change I can still write 

```
class IdentifyProtocol(var idMessage: IdentifyOuterClass.Identify? = null) :
    ProtocolHandler<IdentifyController>() {
     ...
}
```

it will compile and, up to a point, it'll even run. But then it'll throw an exception and it's not clear why. I understand your point about not want to accomodate everything in the Stream class, but that isn't what's happening here. We know, because we already have implementations or they're described in libp2p specs, that some protocols are defined using protobufs. It makes sense to me that would should handle those protobuf-using protocols in the lowest friction way possible. That's what I'm trying to do.